File: 2019/day13/cpp/src/pong.cpp
#include "pong.hpp"
#include <signal.h>
#include <stdlib.h>
#include <stdio.h>
#include <tuple>
#include <iostream>
#include "ncurses.h"

// Initialize the intcode
Pong::Pong(std::string file_name, bool play) : m_intcode(),m_map(), 
    m_ball({0,0,0,0}), m_pad({0,0}), m_map_rendered(false),
    m_ball_rendered(false), m_score(0) {

  m_intcode = Intcode(file_name,{},false,false);

  if (play) {
    m_intcode.m_program[0] = 2;
  }

}

// Generate the map
int Pong::render_map() {

  while (true) {
    
    auto output = m_intcode();
    int x = std::get<1>(output);

    // If terminated
    if (std::get<0>(output)) {
      int nTwos = 0;
      for (auto &m : m_map) {
        if (m.second==2) { nTwos++; }
      }
      return nTwos;
    }

    // Update map
    auto output_y = m_intcode();
    int y = std::get<1>(output_y);

    auto output_t = m_intcode();
    int t = std::get<1>(output_t);

    m_map[std::pair<int,int>(y,x)] = t;
    }
}

// Select an input
int Pong::bot() {
  if (m_ball[3] == 1 && m_pad[1] < m_ball[1]) {
    return 1;
  }
  if (m_ball[3] == -1 && m_pad[1] > m_ball[1]) {
    return -1;
  }
  return 0;
}

// Run the game
int Pong::play() {

  signal(SIGINT, catch_ctrlc);
  initscr();
  
  while(true) {
    auto output = m_intcode();
    auto x = std::get<1>(output);

    // If terminated
    if (std::get<0>(output)) {
      endwin();
      return m_score;  
    }
    
    auto output_y = m_intcode();
    auto output_t = m_intcode();
    int y = std::get<1>(output_y);
    int t = std::get<1>(output_t);

    // Update rules
    if (!(x==-1 && y==0)) {
      m_map[std::pair<int,int>(y,x)] = t;

      if (t == 4) {
        m_ball_rendered = true;
        m_ball = std::vector<int64_t>{y,x,y-m_ball[0],x-m_ball[1]};
      }
      if (t == 3) {
        m_pad = std::vector<int64_t>{y,x};
      }
    } else {
      m_map_rendered = true;
      m_score = t;
      mvprintw(0, 0, "Score: %d\n",m_score);
      refresh();
    }

    if (m_map_rendered && m_ball_rendered) {
      render();
      m_intcode.m_input.push_back(bot());
      m_ball_rendered = false;
    }
  }
}

// Print the map
void Pong::render() {

  int height, width;
  getmaxyx(stdscr, height, width);

  int max_x(WINT_MIN);
  int max_y(WINT_MIN);

  for (auto &m : m_map)
  {
    max_x = (max_x < m.first.first) ? m.first.first : max_x;
    max_y = (max_y < m.first.second) ? m.first.second : max_y;
  }
  for (int row = 0; row < max_x + 1; row++)
  {
    std::string str = "";
    for (int col = 0; col < max_y + 1; col++) {
      int mp = m_map[std::pair<int, int>(row, col)];
      if (mp == 4) {
        str += "o";
      } else if ( mp == 3) {
        str += "=";
      } else if (mp == 2) {
        str += "x";
      } else if (mp == 1) {
        str += "|";
      } else {
        str += " ";
      }
      if (col > width - 3) {
        break;
      }
      mvprintw(row+1,0,str.c_str());
    }
    if (row+1 > height -3) {
      break;
    }
  }
  refresh();
}

// Catch a ctrl-c event
void Pong::catch_ctrlc(int s) {
  endwin();
  printf("Caught signal %d\n",s);
  exit(1);
}

File: 2019/day13/python/pong.py
from os.path import dirname
from os.path import realpath
from os.path import join
from collections import defaultdict
import curses
import sys

sys.path.append(join(dirname(realpath(__file__)), *["..", ".."]))
from day5.python.intcode import Intcode


class Pong:
    """Plays pong from intcode computer instructions."""

    def __init__(self, fileName, play=False):
        """Initialize the pong class with the intcode computer."""

        self.map = defaultdict(lambda: -1)
        self.ic = Intcode(fileName, verbose=False, reset=False)
        if play:
            self.ic.memory[0] = 2

        self.ball, self.pad = [0] * 4, [0] * 2
        self.map_rendered, self.ball_rendered = False, False
        self.score = 0

    def render_map(self):
        """Renders the map given the original instruction and returns the number of
        destroyable blocks."""
        while True:
            cond, x = self.ic()
            if cond:
                # Number of elements in dict that has value 2
                return len([k for k, v in self.map.items() if v == 2])

            _, y = self.ic()
            _, t = self.ic()
            self.map[(y, x)] = t

    def bot(self):
        """Bot that plays the pong game with simple logic"""
        if self.ball[3] == 1 and self.pad[1] < self.ball[1]:
            return 1
        if self.ball[3] == -1 and self.pad[1] > self.ball[1]:
            return -1
        return 0

    def play(self):
        """Play a game of pong with the bot as player."""
        # Screen for rendering the game
        screen = curses.initscr()
        curses.cbreak()
        try:
            while True:
                cond, x = self.ic()
                if cond:
                    break
                _, y = self.ic()
                _, t = self.ic()

                # If not a print instruction update map/ball/pad
                if not (x == -1 and y == 0):
                    self.map[(y, x)] = t
                    if t == 4:
                        self.ball_rendered = True
                        self.ball = (y, x, y - self.ball[0], x - self.ball[1])
                    if t == 3:
                        self.pad = (y, x)
                else:
                    self.map_rendered = True
                    self.score = t
                    screen.addstr(0, 0, "Score: %s " % self.score)

                # If map is rendered, and ball is rendered again apply input
                if self.map_rendered and self.ball_rendered:
                    self.render(screen)
                    self.ic.input.append(self.bot())
                    self.ball_rendered = False

        # Exit screen
        finally:
            curses.echo()
            curses.nocbreak()
            curses.endwin()
            return self.score

    def render(self, screen):
        """Go through the game map and print the walls/obstacles/pad/ball."""
        try:
            height, width = screen.getmaxyx()
            y_max, x_max = map(max, zip(*self.map.keys()))

            mp = {4: "o", 3: "=", 2: "x", 1: "|"}
            for row in range(y_max + 1):
                msg = []
                for k in range(x_max + 1):

                    if self.map[row, k] in mp:
                        msg.append(mp[self.map[row, k]])
                    else:
                        msg.append(" ")
                    if k > width - 3:
                        break
                screen.addstr(row + 1, 0, "".join(msg))
                if row > height - 3:
                    break
            screen.refresh()
        except KeyboardInterrupt:
            curses.echo()
            curses.nocbreak()
            curses.endwin()
            exit(0)
